File: src/io/github/francoiscambell/clionarduinoplugin/CMakeListsEditor.java
package io.github.francoiscambell.clionarduinoplugin;

import com.intellij.openapi.application.*;
import com.intellij.openapi.command.*;
import com.intellij.openapi.editor.*;
import com.intellij.openapi.fileEditor.*;
import com.intellij.openapi.vfs.*;

import java.util.*;

/**
 * Created by francois on 15-08-04.
 */
public class CMakeListsEditor {
    private static Map<VirtualFile, CMakeListsEditor> INSTANCES = new WeakHashMap<VirtualFile, CMakeListsEditor>();
    private VirtualFile cMakeListsVirtualFile;

    private CMakeListsEditor(VirtualFile cMakeLists) {
        this.cMakeListsVirtualFile = cMakeLists;
    }

    public static CMakeListsEditor getInstance(VirtualFile cMakeLists) {
        if (!INSTANCES.containsKey(cMakeLists)) {
            INSTANCES.put(cMakeLists, new CMakeListsEditor(cMakeLists));
        }
        return INSTANCES.get(cMakeLists);
    }

    private VirtualFile getCMakeListsVirtualFile() {
        return cMakeListsVirtualFile;
    }

    private Document getCMakeListsDocument() {
        return FileDocumentManager.getInstance().getDocument(getCMakeListsVirtualFile());
    }

    public void clear() {
        final Document cMakeLists = getCMakeListsDocument();
        ApplicationManager.getApplication().runWriteAction(new Runnable() {
            @Override
            public void run() {
                CommandProcessor.getInstance().executeCommand(null, new Runnable() {
                    @Override
                    public void run() {
                        cMakeLists.deleteString(0, cMakeLists.getTextLength());
                        FileDocumentManager.getInstance().saveDocument(cMakeLists);
                    }
                }, null, null, cMakeLists);
            }
        });
    }

    public void addLine(int line, final String text) {
        final Document cMakeLists = getCMakeListsDocument();
        final int lineEndOffset = cMakeLists.getLineEndOffset(line);

        ApplicationManager.getApplication().runWriteAction(new Runnable() {
            @Override
            public void run() {
                CommandProcessor.getInstance().executeCommand(null, new Runnable() {
                    @Override
                    public void run() {
                        cMakeLists.insertString(lineEndOffset, text + "\n");
                        FileDocumentManager.getInstance().saveDocument(cMakeLists);
                    }
                }, null, null, cMakeLists);
            }
        });
    }

    public void appendLine(String text) {
        int lastLine = getCMakeListsDocument().getLineCount();
        if (lastLine != 0) {
            lastLine--;
        }
        addLine(lastLine, text);
    }

    public void blankLine() {
        appendLine("");
    }

    public void minVersion(String minVersion) {
        method("cmake_minimum_required", "VERSION", minVersion);
    }

    public void set(String var, String value) {
        method("set", var, value);
    }

    public void project(String projectName) {
        method("project", projectName);
    }

    public void method(String methodName, String... args) {
        StringBuilder builder = new StringBuilder(methodName);
        builder.append('(');
        for (String arg : args) {
            builder.append(arg);
            builder.append(' ');
        }
        if (args.length > 0) {
            builder.deleteCharAt(builder.length() - 1);
        }
        builder.append(')');
        appendLine(builder.toString());
    }
}

File: src/io/github/francoiscambell/clionarduinoplugin/CMakeListsEditor.kt
package io.github.francoiscambell.clionarduinoplugin

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.vfs.VirtualFile
import java.util.*

/**
 * Created by francois on 15-08-04.
 */
class CMakeListsEditor private constructor(private val cMakeListsVirtualFile: VirtualFile) {

    private val cMakeListsDocument: Document
        get() = FileDocumentManager.getInstance().getDocument(cMakeListsVirtualFile)

    fun clear() {
        val cMakeLists = cMakeListsDocument
        ApplicationManager.getApplication().runWriteAction {
            CommandProcessor.getInstance().executeCommand(null, {
                cMakeLists.deleteString(0, cMakeLists.textLength)
                FileDocumentManager.getInstance().saveDocument(cMakeLists)
            }, null, null, cMakeLists)
        }
    }

    fun addLine(line: Int, text: String) {
        val cMakeLists = cMakeListsDocument
        val lineEndOffset = cMakeLists.getLineEndOffset(line)

        ApplicationManager.getApplication().runWriteAction {
            CommandProcessor.getInstance().executeCommand(null, {
                cMakeLists.insertString(lineEndOffset, text + "\n")
                FileDocumentManager.getInstance().saveDocument(cMakeLists)
            }, null, null, cMakeLists)
        }
    }

    fun appendLine(text: String) {
        var lastLine = cMakeListsDocument.lineCount
        if (lastLine != 0) {
            lastLine--
        }
        addLine(lastLine, text)
    }

    fun blankLine() {
        appendLine("")
    }

    fun minVersion(minVersion: String) {
        method("cmake_minimum_required", "VERSION", minVersion)
    }

    operator fun set(`var`: String, value: String) {
        method("set", `var`, value)
    }

    fun project(projectName: String) {
        method("project", projectName)
    }

    fun method(methodName: String, vararg args: String) {
        val builder = StringBuilder(methodName)
        builder.append('(')
        for (arg in args) {
            builder.append(arg)
            builder.append(' ')
        }
        if (args.size > 0) {
            builder.deleteCharAt(builder.length - 1)
        }
        builder.append(')')
        appendLine(builder.toString())
    }

    companion object {
        private val INSTANCES = WeakHashMap<VirtualFile, CMakeListsEditor>()

        fun getInstance(cMakeLists: VirtualFile): CMakeListsEditor {
            if (!INSTANCES.containsKey(cMakeLists)) {
                INSTANCES.put(cMakeLists, CMakeListsEditor(cMakeLists))
            }
            return INSTANCES[cMakeLists]
        }
    }
}
